src/Castle.Core/DynamicProxy/ProxyUtil.cs

```diff
@@ -26,11 +26,83 @@ namespace Castle.DynamicProxy
 #endif
 
 	using Castle.Core.Internal;
+	using Castle.DynamicProxy.Generators;
 
 	public static class ProxyUtil
 	{
 		private static readonly SynchronizedDictionary<Assembly, bool> internalsVisibleToDynamicProxy = new SynchronizedDictionary<Assembly, bool>();
 
+		/// <summary>
+		///   Attempts to create a delegate of the specified type to a suitable `Invoke` method
+		///   on the given <paramref name="proxy"/> instance.
+		/// </summary>
+		/// <param name="proxy">The proxy instance to which the delegate should be bound.</param>
+		/// <typeparam name="TDelegate">The type of delegate that should be created.</typeparam>
+		/// <param name="delegate">The produced delegate if the method succeeds; otherwise <see langword="null"/>.</param>
+		/// <returns><see langword="true"/> if the method succeeds; otherwise <see langword="false"/>.</returns>
+		public static bool TryCreateDelegateToMixin<TDelegate>(object proxy, out TDelegate @delegate)
+		{
+			if (TryCreateDelegateToMixin(proxy, typeof(TDelegate), out var d))
+			{
+				@delegate = (TDelegate)(object)d;
```

Comment:
At some point (hopefully with the next major version of Castle.Core), we would want to add a generic type constraint `where TDelegate : MulticastDelegate` to the method declaration to make this kind of casting redundant. For the moment, I've opted not to because the Roslyn compiler version shipping with Mono might be too old to support this. Even if it did, we'd probably need to add a few more of these constraints in some other places; perhaps better to do it all in one go.

Reply:
Makes sense, good call.